tests/unit/generator_test.cc

```diff
@@ -0,0 +1,171 @@
+fibonacci_sequence(unsigned count,
+                   do_suspend suspend) {
+    auto a = 0, b = 1;
+    for (unsigned i = 0; i < count; ++i) {
+        if (std::numeric_limits<decltype(a)>::max() - a < b) {
+            throw std::out_of_range(
+                fmt::format("fibonacci[{}] is greater than the largest value of int", i));
+        }
+        if (suspend) {
+            co_await yield();
+        }
+        co_yield std::exchange(a, std::exchange(b, a + b));
+    }
+}
+
+seastar::future<> test_async_generator_drained(do_suspend suspend) {
+    auto expected_fibs = {0, 1, 1, 2};
+    auto expected_fib = std::begin(expected_fibs);
+
+    auto actual_fibs = fibonacci_sequence(std::size(expected_fibs), suspend);
+    auto actual_fib = co_await actual_fibs.begin();
+
+    for (; actual_fib != actual_fibs.end(); co_await ++actual_fib) {
+        BOOST_REQUIRE(expected_fib != std::end(expected_fibs));
+        BOOST_REQUIRE_EQUAL(*actual_fib, *expected_fib);
+        ++expected_fib;
+    }
+    BOOST_REQUIRE(actual_fib == actual_fibs.end());
+}
+
+SEASTAR_TEST_CASE(test_async_generator_drained_with_suspend) {
+    return test_async_generator_drained(do_suspend::yes);
+}
+
+SEASTAR_TEST_CASE(test_async_generator_drained_without_suspend) {
+    return test_async_generator_drained(do_suspend::no);
+}
+
+seastar::future<> test_async_generator_not_drained(do_suspend suspend) {
+    auto fib = fibonacci_sequence(42, suspend);
+    auto actual_fib = co_await fib.begin();
+    BOOST_REQUIRE_EQUAL(*actual_fib, 0);
+}
+
+SEASTAR_TEST_CASE(test_async_generator_not_drained_with_suspend) {
+    return test_async_generator_not_drained(do_suspend::yes);
+}
+
+SEASTAR_TEST_CASE(test_async_generator_not_drained_without_suspend) {
+    return test_async_generator_not_drained(do_suspend::no);
+}
+
+struct counter_t {
+    int n;
+    int* count;
+    counter_t(counter_t&& other) noexcept
+        : n{std::exchange(other.n, -1)},
+          count{std::exchange(other.count, nullptr)}
+    {}
+    counter_t(int n, int* count) noexcept
+        : n{n}, count{count} {
+        ++(*count);
+    }
+    ~counter_t() noexcept {
+        if (count) {
+            --(*count);
+        }
+    }
+};
+
+std::ostream& operator<<(std::ostream& os, const counter_t& c) {
+    return os << c.n;
+}
+
+coroutine::experimental::async_generator<counter_t>
+fiddle(int n, int* total) {
+    int i = 0;
+    while (true) {
+        if (i++ == n) {
+            throw std::invalid_argument("Eureka from generator!");
+        }
+        co_yield counter_t{i, total};
+    }
+}
+
+SEASTAR_TEST_CASE(test_async_generator_throws_from_generator) {
+    int total = 0;
+    auto count_to = [total=&total](unsigned n) -> seastar::future<> {
+        auto count = fiddle(n, total);
+        auto it = co_await count.begin();
+        for (unsigned i = 0; i < 2 * n; i++) {
+            co_await ++it;
+        }
+    };
+    co_await count_to(42).then_wrapped([&total] (auto f) {
```

Comment:
Instead of `then_wrapped` you could use coroutine::as_future() to keep this code continuations-free.

Reply:
thank you. TIL!